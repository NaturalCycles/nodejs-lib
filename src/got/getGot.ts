import { URL } from 'url'
import { _since } from '@naturalcycles/js-lib'
import got, { AfterResponseHook, BeforeErrorHook, BeforeRequestHook, Got, HTTPError } from 'got'
import { inspectAny } from '..'
import { dimGrey, grey, red, yellow } from '../colors'
import { GetGotOptions, GotRequestContext } from './got.model'

/**
 * Returns instance of Got with "reasonable defaults":
 *
 * 1. Error handler hook that prints helpful errors.
 * 2. Hooks that log start/end of request (optional, false by default).
 * 3. Reasonable defaults(tm), e.g non-infinite Timeout
 */
export function getGot(opt: GetGotOptions = {}): Got {
  opt.logger ||= console

  return got.extend({
    // Most-important is to set to anything non-empty (so, requests don't "hang" by default).
    // Should be long enough to handle for slow responses from scaled cloud APIs in times of spikes
    // Ideally should be LESS than default Request timeout in backend-lib (so, it has a chance to error
    // before server times out with 503).
    timeout: 90_000,
    ...opt,
    hooks: {
      ...opt.hooks,
      beforeError: [
        ...(opt.hooks?.beforeError || []),
        // User hooks go BEFORE
        gotErrorHook(opt),
      ],
      beforeRequest: [
        gotBeforeRequestHook(opt),
        // User hooks go AFTER
        ...(opt.hooks?.beforeRequest || []),
      ],
      afterResponse: [
        ...(opt.hooks?.afterResponse || []),
        // User hooks go BEFORE
        gotAfterResponseHook(opt),
      ],
    },
  })
}

/**
 * Without this hook (default behaviour):
 *
 * HTTPError: Response code 422 (Unprocessable Entity)
 *  at EventEmitter.<anonymous> (.../node_modules/got/dist/source/as-promise.js:118:31)
 *  at processTicksAndRejections (internal/process/task_queues.js:97:5) {
 *  name: 'HTTPError'
 *
 *
 * With this hook:
 *
 * HTTPError 422 GET http://a.com/err?q=1 in 8 ms
 * {
 *   message: 'Reference already exists',
 *   documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
 * }
 *
 * Features:
 * 1. Includes original method and URL (including e.g searchParams) in the error message.
 * 2. Includes response.body in the error message (limited length).
 * 3. Auto-detects and parses JSON response body (limited length).
 * 4. Includes time spent (gotBeforeRequestHook must also be enabled).
 * UPD: excluded now to allow automatic Sentry error grouping
 */
function gotErrorHook(opt: GetGotOptions = {}): BeforeErrorHook {
  const { maxResponseLength = 10_000 } = opt

  return err => {
    if (err instanceof HTTPError) {
      const { statusCode } = err.response
      const { method, url, prefixUrl } = err.options
      const shortUrl = getShortUrl(opt, url, prefixUrl)
      // const { started } = context as GotRequestContext

      const body = inspectAny(err.response.body, {
        maxLen: maxResponseLength,
        colors: false,
      })

      // timings are not part of err.message to allow automatic error grouping in Sentry
      err.message = [[statusCode, method, shortUrl].filter(Boolean).join(' '), body]
        .filter(Boolean)
        .join('\n')
    }

    return err
  }
}

function gotBeforeRequestHook(opt: GetGotOptions): BeforeRequestHook {
  return options => {
    options.context = {
      ...options.context,
      started: Date.now(),
    } as GotRequestContext

    if (opt.logStart) {
      const shortUrl = getShortUrl(opt, options.url, options.prefixUrl)
      opt.logger!.log([dimGrey(' >>'), dimGrey(options.method), grey(shortUrl)].join(' '))
    }
  }
}

function gotAfterResponseHook(opt: GetGotOptions = {}): AfterResponseHook {
  return resp => {
    const success = resp.statusCode >= 200 && resp.statusCode < 400

    if (opt.logFinished) {
      const { started } = resp.request.options.context as GotRequestContext
      const { url, prefixUrl, method } = resp.request.options
      const shortUrl = getShortUrl(opt, url, prefixUrl)

      opt.logger!.log(
        [
          dimGrey(' <<'),
          coloredHttpCode(resp.statusCode),
          dimGrey(method),
          grey(shortUrl),
          started && dimGrey('in ' + _since(started)),
        ]
          .filter(Boolean)
          .join(' '),
      )
      // console.log(`afterResp! ${resp.request.options.method} ${resp.url}`, { context: resp.request.options.context })
    }

    // Error responses are not logged, cause they're included in Error message already
    if (opt.logResponse && success) {
      opt.logger!.log(inspectAny(resp.body, { maxLen: opt.maxResponseLength }))
    }

    return resp
  }
}

function coloredHttpCode(statusCode: number): string {
  if (statusCode < 400) return dimGrey(statusCode) // default
  if (statusCode < 500) return yellow(statusCode)
  return red(statusCode)
}

function getShortUrl(opt: GetGotOptions, url: URL, prefixUrl?: string): string {
  let shortUrl = url.toString()

  if (opt.logWithSearchParams === false) {
    shortUrl = shortUrl.split('?')[0]!
  }

  if (opt.logWithPrefixUrl === false && prefixUrl && shortUrl.startsWith(prefixUrl)) {
    shortUrl = shortUrl.slice(prefixUrl.length)
  }

  return shortUrl
}
