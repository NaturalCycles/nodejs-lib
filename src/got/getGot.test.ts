import { mockTime } from '@naturalcycles/dev-lib/dist/testing'
import { _anyToErrorObject, _range, pTry } from '@naturalcycles/js-lib'
import nock = require('nock')
import { arraySchema, getValidationResult, HTTPError, integerSchema, objectSchema } from '..'
import { getGot } from './getGot'

beforeAll(() => {
  mockTime()
})

afterAll(() => {
  nock.cleanAll()
  nock.restore() // prevents jest memory leak
})

const okResp = { ok: 1 }
const errResp = {
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference',
}

const items = _range(100).map(id => ({ id }))
const itemsSchema = arraySchema(objectSchema({ id: integerSchema.max(98) }))
const { error: validationError } = getValidationResult(items, itemsSchema)
validationError!.data = { ...validationError!.data, httpStatusCode: 400 }
const joiErrorBackendResp = { error: _anyToErrorObject(validationError) }

nock(/.*/)
  .persist()
  .get(/.*/)
  .reply((uri, _body: any) => {
    // console.log({uri, _body})

    if (uri.includes('backendErr')) {
      return [400, joiErrorBackendResp]
    }

    if (uri.includes('err')) {
      return [422, errResp]
    }

    return [200, okResp]
  })

const got = getGot({
  logStart: true,
  logFinished: true,
  logResponse: true,
})

test('gotErrorHook', async () => {
  expect(await got.get('http://a.com').json()).toEqual(okResp)

  await expect(
    got.get('http://a.com/err', {
      searchParams: { q: 1 },
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('backend error', async () => {
  const [err] = await pTry(
    got.get('http://a.com/backendErr', {
      searchParams: { q: 1 },
    }),
  )

  expect(err).toBeInstanceOf(HTTPError)
  expect(err).toMatchSnapshot()
})

test('logWithSearchParams=false', async () => {
  const g = getGot({
    logStart: true,
    logFinished: true,
    logResponse: true,
    logWithSearchParams: false,
  })

  await expect(
    g.get('http://a.com/err', {
      searchParams: { q: 1 },
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('logWithPrefixUrl=false', async () => {
  const g = getGot({
    logStart: true,
    logFinished: true,
    logResponse: true,
    prefixUrl: 'http://a.com',
    logWithPrefixUrl: false,
  })

  await expect(
    g.get('err', {
      searchParams: { q: 1 },
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})
