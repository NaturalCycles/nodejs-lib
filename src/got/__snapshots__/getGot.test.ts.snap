// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backend error 1`] = `
[HTTPError: 400 GET http://a.com/backendErr?q=1
JoiValidationError: [
  {
    "id": 0
  },
  {
    "id": 1
  },
  {
    "id": 2
  },
  {
    "id": 3
  },
  {
    "id": 4
  },
  {
    "id": 5
  },
  {
    "id": 6
  },
  {
    "id": 7
  },
  {
    "id": 8
  },
  {
    "id": 9
  },
  {
    "id": 10
  },
  {
    "id": 11
  },
  {
    "id": 12
  },
  {
    "id": 13
  },
  {
    "id": 14
  },
  {
    "id": 15
  },
  {
    "id": 16
  },
  {
    "id": 17
  },
  {
    "id": 18
  },
  {
    "id": 19
  },
  {
    "id": 20
  },
  {
    "id": 21
  },
  {
    "id": 22
  },
  {
    "id": 23
  },
  {
    "id": 24
  },
  {
    "id": 25
  },
  {
    "id": 26
  },
  {
    "id": 27
  },
  {
    "id": 28
  },
  {
    "id": 29
  },
  {
    "id": 30
  },
  {
    "id": 31
  },
  {
    "id": 32
  },
  {
    "id": 33
  },
  {
    "id": 34
  },
  {
    "id": 35
  },
  {
    "id": 36
  },
  {
    "id": 37
  },
  {
    "id": 38
  },
  {
    "id": 39
  },
  {
    "id": 40
  },
  {
    "id": 41
  },
  {
    "id": 42
  },
  {
    "id": 43
  },
  {
    "id": 44
  },
  {
    "id": 45
  },
  {
    "id": 46
  },
  {
    "id": 47
  },
  {
    "id": 48
  },
  {
    "id": 49
  },
  {
    "id": 50
  },
  {
    "id": 51
  },
  {
    "id": 52
  },
  {
    "id": 53
  },
  {
    "id": 54
  },
  {
    "id": 55
  },
  {
    "id": 56
  },
  {
    "id": 57
  },
  {
    "id": 58
  },
  {
    "id": 59
  },
  {
    "id": 60
  },
  {
    "id": 61
  },
  {
    "id": 62
  },
  {
    "id": 63
  },
  {
    "id": 64
  },
  {
    "id": 65
  },
  {
    "id": 66
  },
  {
    "id": 67
  },
  {
    "id": 68
  },
  {
    "id": 69
  },
  {
    "id": 70
  },
  {
    "id": 71
  },
  {
    "id": 72
  },
  {
    "id": 73
  },
  {
    "id": 74
  },
  {
    "id": 75
  },
  {
    "id": 76
  },
  {
    "id": 77
  },
  {
    "id": 78
  },
  {
    "id": 79
  },
  {
    "id": 80
  },
  {
    "id": 81
  },
  {
    "id": 82
  },
  {
    "id": 83
  },
  {
    "id": 84
  },
  {
    "id": 85
  },
  {
    "id": 86
  },
  {
    "id": 87
  },
  {
    "id": 88
  },
  {
    "id": 89
  },
  {
    "id": 90
  },
  {
  
... 3 KB message truncated
"[99].id" must be less than or equal to 98 @ .99.id]
`;

exports[`gotErrorHook 1`] = `
"422 GET http://a.com/err?q=1
{
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
}"
`;

exports[`logWithPrefixUrl=false 1`] = `
"422 GET err?q=1
{
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
}"
`;

exports[`logWithSearchParams=false 1`] = `
"422 GET http://a.com/err
{
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
}"
`;
